File: OOPRecords.Model/SetRepository.cs
using NakedObjects;
using System.Linq;

namespace OOPRecords.Model
{
    public class SetRepository
    {
        public IDomainObjectContainer Container { set; protected get; }

        public TeachingSet CreateNewSet()
        {
            return Container.NewTransientInstance<TeachingSet>();
        }

        public IQueryable<TeachingSet> ListSets([Optionally] Subject subject, [Optionally] int? yearGroup)
        {
            var sets = Container.Instances<TeachingSet>();
            if (subject != null)
            {
                int id = subject.Id;
                sets = sets.Where(s => s.Subject.Id == id);
            }
            if (yearGroup != null)
            {
                sets = sets.Where(s => s.YearGroup == yearGroup.Value);
            }
            return sets.OrderBy(s => s.YearGroup).ThenBy(s => s.Subject.Name);
        }
    }
}

File: OOPRecords.ModelVB/SetRepository.vb
Imports NakedObjects

Public Class SetRepository
    Private _container As IDomainObjectContainer
    Public WriteOnly Property Container() As IDomainObjectContainer
        Set(ByVal value As IDomainObjectContainer)
            _container = value
        End Set
    End Property

    Public Function CreateNewSet() As TeachingSet
        Return _container.NewTransientInstance(Of TeachingSet)()
    End Function

    Public Function ListSets(<Optionally> ByVal subject As Subject, <Optionally> ByVal yearGroup As Integer?) As IQueryable(Of TeachingSet)
        Dim sets = _container.Instances(Of TeachingSet)()

        If subject IsNot Nothing Then
            Dim id As Integer = subject.Id
            sets = sets.Where(Function(s) s.Subject.Id = id)
        End If

        If yearGroup IsNot Nothing Then
            sets = sets.Where(Function(s) s.YearGroup = yearGroup.Value)
        End If

        Return sets.OrderBy(Function(s) s.YearGroup).ThenBy(Function(s) s.Subject.Name)
    End Function
End Class